server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/GetFieldMappingsResponse.java

```diff
@@ -93,9 +102,75 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         return builder;
     }
 
+    private static ObjectParser createParser(Map<String, Map<String, Map<String, FieldMappingMetaData>>> mappings) {
+        final ObjectParser<Map<String, Map<String, Map<String, FieldMappingMetaData>>>, String> parser =
+            new ObjectParser<>(MAPPINGS.getPreferredName(), true, () -> mappings);
+
+        parser.declareField((p, b, index) -> {
+            final Map<String, Object> map = p.map();
+            final Map<String, Map<String, FieldMappingMetaData>> typeMappings = new HashMap<>();
+            for (Map.Entry<String, Object> typeObjectEntry : map.entrySet()) {
+                final Map<String, FieldMappingMetaData> fieldsMapping = new HashMap<>();
+                typeMappings.put(typeObjectEntry.getKey(), fieldsMapping);
+
+                final Object o1 = typeObjectEntry.getValue();
+                if (!(o1 instanceof Map)) {
+                    throw new ParsingException(p.getTokenLocation(),
+                        "Nested type mapping at " + index + "." + typeObjectEntry.getKey() + " is not found");
+                }
+                final Map<String, Object> map2 = (Map) o1;
+                for (Map.Entry<String, Object> e : map2.entrySet()) {
+                    final Object o2 = e.getValue();
+                    if (!(o2 instanceof Map)) {
+                        throw new ParsingException(p.getTokenLocation(),
+                            "Nested field mapping at " + index + "." + typeObjectEntry.getKey() + "." + e.getKey()
+                                + " is not found");
+                    }
+                    final Map<String, Object> map3 = (Map) o2;
+
+                    final String fullName = (String) map3.get(FieldMappingMetaData.FULL_NAME.getPreferredName());
+                    final XContentBuilder jsonBuilder = jsonBuilder();
+                    final Map<String, ?> values = (Map<String, ?>) map3.get(FieldMappingMetaData.MAPPING.getPreferredName());
+                    jsonBuilder.map(values);
+                    final BytesReference source = BytesReference.bytes(jsonBuilder);
+
+                    final FieldMappingMetaData metaData = new FieldMappingMetaData(fullName, source);
+                    fieldsMapping.put(e.getKey(), metaData);
+                }
+
+            }
+            mappings.put(index, typeMappings);
+        }, MAPPINGS, ObjectParser.ValueType.OBJECT);
+
+        return parser;
+
+    }
+
+    public static GetFieldMappingsResponse fromXContent(XContentParser parser) throws IOException {
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser::getTokenLocation);
+
+        final Map<String, Map<String, Map<String, FieldMappingMetaData>>> mappings = new HashMap<>();
+        final ObjectParser objectParser = createParser(mappings);
```

Comment:
It is much more normal to create the objectParser statically. I see that you did it this way so that you could collect the results into something, but that isn't really what object parser is *for*. I mean, you can do it by passing the `mappings` map as the context to the parse method and throwing out the return value, but that is weird. Why not have the `ObjectParser` return the `typeMappings`?

Reply:
even better - agree